docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,813 @@
+privilege escalation (e.g using [distroless](https://github.com/GoogleContainerTools/distroless) base images) and
+the Pod spec in the Deployment manifest should enforce security best practices (e.g. do not use privileged pods etc.).
+
+####  Alternative deployments methods
+Alternative deployment methods can be used given that the requirement about the HTTP endpoint accessibility
+is satisfied, like e.g.
+
+- deploying the HTTP Server as a part of another component, e.g. a controller;
+- deploying the HTTP Server outside the Management Cluster.
+
+In these cases above recommendations about availability and identity and access management apply as well.
+
+### Registering Runtime Extensions
+
+_Important! Cluster administrators should carefully vet any Runtime Extension registration, thus preventing
+malicious components from being added to the system._
+
+_Creating RuntimeExtensionConfiguration will be allowed only if the Runtime Extension feature flag is set to true._
+
+By registering a Runtime Extension the Cluster API Runtime becomes aware of a Runtime Extension implementing a
+Runtime Hook, and as a consequence the runtime starts calling the extension at well-defined moments of the
+workload cluster’s lifecycle.
+
+This process has many similarities with registering dynamic webhooks in Kubernetes, but in this case it is required
+to use the following CR kind introduced by this Proposal:
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: RuntimeExtensionConfiguration
+metadata:
+  name: "my-amazing-product-runtime-extension"
+webhooks:
+# Name should be fully qualified and unique in the Cluster, thus usage of sub domains, version qualifiers is recommended.
+- name: "my-amazing-runtime-extension.v1.5.panda.com"
+  # List of group/version/hook the RuntimeExtension implements. Required
+  operations:
+  - apiVersion: "cluster.runtime.cluster.x-k8s.io/v1alpha2"
+    hook: "beforeUpgrade"
+  - apiVersion: "cluster.runtime.cluster.x-k8s.io/v1alpha2"
+    hook: "afterUpgrade"
+  #ClientConfig defines how to communicate with the RuntimeExtension. Required
+  clientConfig:
+    #`url` gives the location of the RuntimeExtension, in standard URL form (`scheme://host:port/path`). Exactly one of `url` or `service` must be specified.
+    url: "..."
+    service:
+      namespace: "example-namespace"
+      name: "example-service"
+      # `path` is an optional path prefix path which can be sent in any request to this service.
+      path: "runtime-extensions/"
+      # If specified, the port on the service that hosts the RuntimeExtension. Default to 443. `port` should be a valid port number (1-65535, inclusive).
+      port: 8082
+    caBundle: "..."
+  # If specified, define the timeout for each RuntimeExtension call to complete (Default 5s, Max 10s).
+  timeoutSeconds: 2
+  # FailurePolicy defines how errors from RuntimeExtension calls are handled - allowed values are Ignore or Fail. Defaults to Fail.
+  failurePolicy: Fail
```

Comment:
should the failure policy be at the group/version/hook level?

Reply:
@enxebre that's a good point.
while investigating this a little bit more, we came to the conclusion that it is convenient to have a RuntimeExtension server to server multiple runtime extensions, and this quickly scales to 10-20 extensions; in order to avoid putting on the cluster administrator the burden to document/maintain such a list, we decided to propose a discovery mechanism, thus shifting this responsibility to the extensions developers. 
I have updated the document accordingly, PTAL